code/controllers/subsystem/gravity.dm

```diff
@@ -0,0 +1,176 @@
+	var/list/currentrun_manual = list()
+	var/recalculation_cost = 0
+	var/do_purge = FALSE
+	var/purge_interval = 600
+	var/purge_tick = 0
+	var/purging = FALSE
+	var/error_no_atom
+	var/list/error_mismatched_area
+	var/list/error_mismatched_turf
+	var/list/error_no_area
+	var/list/error_no_turf
+	var/list/gravgens
+	var/list/purging_atoms
+	var/list/atoms_forced_gravity_processing
+	var/static/list/area_blacklist_typecache = typecacheof(list(/area/lavaland, /area/mine, /area/centcom))
+	var/inited_atoms = 0
+	var/inited_areas = 0
+	var/init_state = 0
+
+/datum/controller/subsystem/gravity/New()
+	NEW_SS_GLOBAL(SSgravity)
+
+/datum/controller/subsystem/gravity/proc/init_lists(full = FALSE)
+	gravgens = list()
+	purging_atoms = list()
+	atoms_forced_gravity_processing = list()
+	if(full)
+		error_no_turf = list()
+		error_no_area = list()
+		error_mismatched_turf = list()
+		error_mismatched_area = list()
+
+/datum/controller/subsystem/gravity/Initialize()
+	init_state = 1
+	init_lists(TRUE)
+	inited_atoms = 0
+	for(var/atom/movable/A in world)
+		A.init_gravity()
+		inited_atoms++
+		CHECK_TICK
+	for(var/area/A in world)
+		A.init_gravity()
+		inited_areas++
+		CHECK_TICK
+	testing("Initialized gravity for [inited_atoms] movable atoms and [inited_areas] areas!")
+	. = ..()
+
+/datum/controller/subsystem/gravity/Recover()
+	init_lists()
+	do_purge = SSgravity.do_purge
+	purge_interval = SSgravity.purge_interval
+	error_no_atom = SSgravity.error_no_atom
+	error_mismatched_area = SSgravity.error_mismatched_area
+	error_mismatched_turf = SSgravity.error_mismatched_turf
+	error_no_area = SSgravity.error_no_area
+	error_no_turf = SSgravity.error_no_turf
+
+/datum/controller/subsystem/gravity/proc/reset_gravity_processing()
+	var/count = 0
+	var/can_fire_old = can_fire
+	can_fire = FALSE
+	while(atoms_forced_gravity_processing.len)
+		var/atom/movable/AM = atoms_forced_gravity_processing[atoms_forced_gravity_processing.len]
+		atoms_forced_gravity_processing.len--
+		if(!istype(AM))
+			SSgravity.error_no_atom++
+			continue
+		else
+			AM.force_gravity_processing = FALSE
+			count++
+		CHECK_TICK
+	LAZYCLEARLIST(atoms_forced_gravity_processing)
+	var/atoms_not_found = "ERROR: NO SUBSYSTEM!"
+	if(SSgravity)
+		atoms_not_found = SSgravity.error_no_atom
+	can_fire = can_fire_old
+	return "[count] atoms purged from forced processing! [atoms_not_found] things found so far that were not atoms!"
+
+/datum/controller/subsystem/gravity/proc/recalculate_atoms()
+	currentrun = list()
+	currentrun_manual = list()
+	var/tempcost = REALTIMEOFDAY
+	for(var/I in sortedAreas)
+		var/area/A = I
+		if(!A.has_gravity && !A.gravity_generator && !A.gravity_overriding)
+			continue
+		if(!A.gravity_direction)	//Right now we don't need this.
+			continue
+		if(is_type_in_typecache(A, area_blacklist_typecache))
+			continue
+		currentrun += A.contents_affected_by_gravity
+	currentrun_manual = atoms_forced_gravity_processing.Copy()
+	recalculation_cost = REALTIMEOFDAY - tempcost
```

Comment:
@optimumtact It is, which is why I'm doing it so I can see if necessary how bad it gets in high lag situations (which have not been tested, as when I'm testing I just fire it out and SDQL every turf/area to random directions and set it on highest processing and see how fast it goes and if it starts blocking up other stuff)